File: src/types/facility.ts
/**
 * Facility Core Service — Domain Types & Data Model
 * 
 * Enterprise-grade Facility Management System for Bashundhara R/A Smart Community.
 * Includes Asset Management, Preventive Maintenance, Work Orders, Utilities,
 * Housekeeping, Vendors, AMC, Compliance, Biomedical, Inventory, Costing & Analytics.
 */

export type AssetCategory =
  | "Generator"
  | "Lift"
  | "Water Pump"
  | "Electrical Panel"
  | "Transformer"
  | "HVAC"
  | "AC"
  | "Water Treatment Equipment"
  | "Fire Safety Equipment"
  | "CCTV"
  | "Access Control Equipment"
  | "Solar Equipment"
  | "Plumbing Equipment"
  | "Cleaning Equipment"
  | "Medical Equipment"
  | "Furniture"
  | "IT Equipment"
  | "Other";

export type AssetStatus =
  | "Active"
  | "Operational"
  | "Under Maintenance"
  | "Damaged"
  | "Inactive"
  | "Disposed"
  | "Retired";

export type AssetCondition = "Excellent" | "Good" | "Fair" | "Poor" | "Critical";

export interface AssetLifecycleEvent {
  id: string;
  assetId: string;
  date: string;
  eventType: "PROCUREMENT" | "RECEIVED" | "INSTALLED" | "OPERATIONAL" | "MAINTENANCE" | "REPAIR" | "REPLACEMENT" | "RETIRED" | "DISPOSED" | "AMC_RENEWED";
  title: string;
  description: string;
  performedBy: string;
  cost?: number;
  workOrderId?: string;
  documentRef?: string;
}

export interface AssetMetric {
  assetId: string;
  runtimeHours: number;
  temperatureC: number;
  vibrationMmS: number;
  pressureBar: number;
  voltageV: number;
  currentA: number;
  waterFlowLpm: number;
  lastUpdated: string;
  healthScore: number; // 0 - 100
  isAnomaly: boolean;
}

export interface FacilityAsset {
  id: string;
  assetCode: string; // e.g. GEN-A-001
  name: string;
  category: AssetCategory;
  subcategory: string;
  manufacturer: string;
  brand: string;
  model: string;
  serialNumber: string;
  purchaseDate: string;
  installationDate: string;
  purchaseCost: number;
  currentValue: number;
  warrantyExpiry: string;
  
  // Location Hierarchy
  community: string;
  block: string;
  road: string;
  buildingId: string;
  buildingName: string;
  floor: string;
  room: string;
  facilityArea: string; // e.g. "Generator Room", "Pump Room", "Lift Shaft"
  
  responsibleDepartment: string;
  responsiblePerson: string;
  vendorId?: string;
  vendorName?: string;
  amcId?: string;
  amcName?: string;
  
  status: AssetStatus;
  condition: AssetCondition;
  expectedLifeYears: number;
  lastMaintenanceDate: string;
  nextMaintenanceDate: string;
  
  qrCodeUrl: string; // Frontend QR placeholder
  documentsCount: number;
  photosCount: number;
  notes: string;
}

/* ---------------------------------------------------- Maintenance & Work Orders */

export type MaintenanceType =
  | "Preventive"
  | "Corrective"
  | "Emergency"
  | "Breakdown"
  | "Inspection"
  | "Predictive";

export type WorkOrderStatus =
  | "New"
  | "Assigned"
  | "Scheduled"
  | "In Progress"
  | "Waiting Parts"
  | "Waiting Vendor"
  | "Completed"
  | "Verified"
  | "Cancelled";

export type WorkOrderPriority = "low" | "normal" | "high" | "critical";

export interface WorkOrderPart {
  id: string;
  partName: string;
  partCode: string;
  quantity: number;
  unitCost: number;
  totalCost: number;
}

export interface FacilityWorkOrder {
  id: string;
  workOrderCode: string; // e.g. WO-FAC-1001
  assetId: string;
  assetName: string;
  assetCode: string;
  location: string;
  buildingId: string;
  issue: string;
  description: string;
  priority: WorkOrderPriority;
  maintenanceType: MaintenanceType;
  requestedBy: string;
  assignedDepartment: string;
  assignedTechnician: string;
  vendorId?: string;
  vendorName?: string;
  scheduledDate: string;
  dueDate: string;
  slaMinutes: number;
  status: WorkOrderStatus;
  
  estimatedCost: number;
  actualCost: number;
  laborCost: number;
  partsCost: number;
  vendorCost: number;
  
  partsUsed: WorkOrderPart[];
  photosCount: number;
  notes: string;
  completionEvidence?: string;
  completedAt?: string;
  verifiedBy?: string;
  
  // Integration refs
  requestId?: string;
  accountsPayableRef?: string;
}

/* --------------------------------------------------- Preventive Maintenance & Checklist */

export interface ChecklistItem {
  id: string;
  task: string;
  result: "Pass" | "Fail" | "N/A" | "Pending";
  notes?: string;
}

export interface PreventiveSchedule {
  id: string;
  scheduleCode: string;
  assetId: string;
  assetName: string;
  assetCode: string;
  maintenanceType: MaintenanceType;
  frequencyDays: number; // e.g. 30, 90, 365
  frequencyLabel: string; // e.g. "Monthly", "Quarterly", "Annual"
  lastCompletedDate: string;
  nextDueDate: string;
  responsiblePerson: string;
  vendorName?: string;
  estimatedCost: number;
  slaMinutes: number;
  status: "Active" | "Paused" | "Overdue" | "Completed";
  checklist: ChecklistItem[];
}

/* ------------------------------------------------------------- Utilities */

export type UtilityType = "Electricity" | "Water" | "Gas" | "Generator Fuel" | "Solar Energy" | "Other";

export interface UtilityMeter {
  id: string;
  meterCode: string; // e.g. MTR-ELEC-01
  utilityType: UtilityType;
  location: string;
  buildingId: string;
  buildingName: string;
  flatId?: string;
  commonAreaName?: string;
  serialNumber: string;
  installationDate: string;
  previousReading: number;
  currentReading: number;
  lastReadingDate: string;
  unit: "kWh" | "m³" | "Liters" | "Other";
  status: "Active" | "Faulty" | "Disconnected";
  ratePerUnit: number; // in BDT
}

export interface UtilityReading {
  id: string;
  meterId: string;
  meterCode: string;
  utilityType: UtilityType;
  buildingName: string;
  previousReading: number;
  currentReading: number;
  consumption: number;
  unit: string;
  readingDate: string;
  readerName: string;
  readingSource: "Manual" | "Scheduled" | "Imported" | "IoT Sensor";
  costBDT: number;
  isAnomaly: boolean;
}

export interface UtilityAlert {
  id: string;
  meterId: string;
  utilityType: UtilityType;
  buildingName: string;
  alertType: "Unusual Consumption" | "Sudden Spike" | "Meter Failure" | "Low Fuel Alert" | "Overdue Reading";
  severity: "high" | "critical" | "medium";
  description: string;
  timestamp: string;
  resolved: boolean;
}

/* ------------------------------------------------------------- Housekeeping */

export interface HousekeepingTask {
  id: string;
  taskCode: string;
  location: string;
  buildingName: string;
  taskName: string;
  frequency: "Daily" | "Shift" | "Weekly" | "Event-based";
  assignedStaff: string;
  startTime: string;
  dueTime: string;
  status: "Pending" | "In Progress" | "Completed" | "Inspected" | "Failed";
  checklist: ChecklistItem[];
  photosCount: number;
  supervisorScore?: "Excellent" | "Good" | "Needs Improvement" | "Failed";
  supervisorNotes?: string;
  reinspectionRequired?: boolean;
}

export interface WasteCollection {
  id: string;
  zone: string;
  buildingName: string;
  collectionSchedule: string;
  collectionStaff: string;
  wasteType: "General Organic" | "Recyclable" | "Hazardous/Medical" | "Bulk Waste";
  pickupStatus: "Scheduled" | "Collected" | "Missed" | "In Progress";
  pickupTime?: string;
  notes?: string;
}

/* ------------------------------------------------------------- Vendors & AMC */

export interface FacilityVendor {
  id: string;
  vendorCode: string;
  companyName: string;
  contactPerson: string;
  phone: string;
  email: string;
  address: string;
  category: string; // e.g. Electrical, Lift, HVAC, Cleaning
  servicesOffered: string[];
  verificationStatus: "Verified" | "Pending" | "Unverified";
  rating: number; // 1 to 5
  status: "Active" | "Suspended" | "Blacklisted";
  activeAMCsCount: number;
  slaComplianceRate: number; // percentage
}

export interface AMCContract {
  id: string;
  contractCode: string; // e.g. AMC-2026-LIFT
  vendorId: string;
  vendorName: string;
  assetId: string;
  assetName: string;
  serviceType: string;
  startDate: string;
  endDate: string;
  contractValueBDT: number;
  paymentTerms: string;
  slaResponseHours: number;
  visitFrequency: "Monthly" | "Bi-Monthly" | "Quarterly" | "Annual";
  includedServices: string;
  excludedServices: string;
  renewalDate: string;
  status: "Draft" | "Active" | "Expiring Soon" | "Expired" | "Cancelled" | "Renewed";
  documentsCount: number;
}

export interface AMCVisit {
  id: string;
  amcId: string;
  contractCode: string;
  vendorName: string;
  assetName: string;
  scheduledDate: string;
  technicianName: string;
  technicianPhone: string;
  securityPassId?: string;
  gateArrivalStatus: "Expected" | "At Gate" | "In Building" | "Completed" | "Cancelled";
  workOrderId?: string;
  serviceReportNotes: string;
  completedAt?: string;
}

/* ------------------------------------------------------------- Compliance & Inspections */

export interface ComplianceRequirement {
  id: string;
  requirementCode: string;
  requirementName: string;
  assetId?: string;
  assetName?: string;
  authorityStandard: string; // e.g. "RAJUK / Fire Service Bangladesh", "BSTI"
  responsibleDepartment: string;
  inspectionFrequencyDays: number;
  lastInspectionDate: string;
  nextInspectionDate: string;
  certificateNumber: string;
  issueDate: string;
  expiryDate: string;
  status: "Compliant" | "Due Soon" | "Overdue" | "Expired" | "Non-Compliant";
  documentsCount: number;
}

export interface FacilityInspection {
  id: string;
  inspectionCode: string;
  type: "Safety" | "Electrical" | "Fire" | "Lift" | "Generator" | "Water Quality" | "Housekeeping" | "Building Envelope";
  targetName: string;
  assignedInspector: string;
  scheduledDate: string;
  status: "Scheduled" | "In Progress" | "Passed" | "Action Required" | "Failed";
  findings: string;
  correctiveWorkOrderId?: string;
  completedDate?: string;
}

/* ------------------------------------------------------------- Biomedical Equipment */

export interface BiomedicalEquipment {
  id: string;
  equipmentCode: string;
  name: string;
  manufacturer: string;
  model: string;
  serialNumber: string;
  clinicLocation: string; // e.g. "Bashundhara Community Clinic - Room 102"
  purchaseDate: string;
  warrantyExpiry: string;
  calibrationFrequencyDays: number;
  lastCalibrationDate: string;
  nextCalibrationDate: string;
  certificationStatus: "Certified" | "Calibration Due" | "Out of Calibration" | "Faulty";
  responsiblePerson: string;
  vendorName: string;
}

export interface CalibrationRecord {
  id: string;
  equipmentId: string;
  equipmentName: string;
  calibrationDate: string;
  calibratedBy: string; // Agency / Vendor
  certificateNumber: string;
  result: "Pass" | "Fail" | "Adjusted";
  costBDT: number;
  nextDue: string;
}

/* ------------------------------------------------------------- Facility Inventory */

export interface FacilityInventoryItem {
  id: string;
  itemCode: string;
  itemName: string;
  category: "Spare Parts" | "Electrical" | "Plumbing" | "Cleaning Chemicals" | "Maintenance Tools" | "Safety Equipment" | "HVAC Parts" | "Generator Parts" | "Consumables";
  quantityInStock: number;
  unit: string;
  reorderLevel: number;
  unitCostBDT: number;
  supplierName: string;
  storageLocation: string; // e.g. "Main Store A - Shelf 3"
  status: "In Stock" | "Low Stock" | "Out of Stock";
}

/* ------------------------------------------------------------- Costing & Budget */

export interface FacilityBudget {
  id: string;
  category: "Maintenance" | "Utilities" | "Housekeeping" | "Security Equipment" | "Asset Replacement" | "AMC" | "Compliance" | "Emergency" | "Capital Expenditure" | "Operational Expenditure";
  allocatedBDT: number;
  actualSpentBDT: number;
  committedBDT: number;
  remainingBDT: number;
  fiscalYear: string;
}

export interface FacilityCostRecord {
  id: string;
  date: string;
  category: string;
  description: string;
  amountBDT: number;
  buildingName: string;
  assetName?: string;
  workOrderRef?: string;
  accountLedgerRef: string;
  status: "Posted" | "Pending Approval";
}

/* ------------------------------------------------------------- Dashboard KPI Summary */

export interface FacilityDashboardSummary {
  totalAssets: number;
  activeAssets: number;
  underMaintenanceAssets: number;
  overdueMaintenanceCount: number;
  openWorkOrders: number;
  criticalIssuesCount: number;
  activeAMCContracts: number;
  amcExpiringSoonCount: number;
  monthlyUtilityCostBDT: number;
  housekeepingPendingCount: number;
  complianceAlertsCount: number;
  expiredCertificatesCount: number;
  monthlyExpenseBDT: number;
  pendingVendorPaymentsBDT: number;
  stockValueBDT: number;
}

File: backend/app/api/v1/endpoints/facility.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.db.session import get_db
from app.models.facility import FacilityAsset, FacilityWorkOrder, UtilityMeter, UtilityReading
from app.schemas.domain import AssetCreate, WorkOrderCreate, UtilityReadingCreate
import uuid

router = APIRouter()

# Assets
@router.get("/assets")
def list_assets(db: Session = Depends(get_db)):
    return db.query(FacilityAsset).all()

@router.post("/assets")
def create_asset(data: AssetCreate, db: Session = Depends(get_db)):
    asset_id = f"AST-{uuid.uuid4().hex[:6].upper()}"
    code = f"{data.category[:3].upper()}-A-{uuid.uuid4().hex[:4].upper()}"
    asset = FacilityAsset(
        id=asset_id,
        asset_code=code,
        name=data.name,
        category=data.category,
        subcategory=data.subcategory,
        manufacturer="Enterprise Manufacturer",
        brand="Standard",
        model="V-2026",
        serial_number=f"SN-{uuid.uuid4().hex[:6].upper()}",
        purchase_date="2024-01-15",
        installation_date="2024-02-01",
        purchase_cost=data.purchaseCost,
        current_value=data.purchaseCost,
        warranty_expiry="2027-12-31",
        community="Bashundhara R/A",
        block="Block A",
        road="Road 5",
        building_id="BLD-004",
        building_name=data.buildingName,
        floor="Ground Floor",
        room=data.room,
        facility_area="Common Area",
        responsible_department="Facility Operations",
        responsible_person="Facility Manager",
        status="Operational",
        condition="Excellent",
        expected_life_years=15,
        last_maintenance_date="2026-08-01",
        next_maintenance_date="2026-09-01",
        qr_code_url=f"/qr-placeholder/{code}",
        notes="Created via API."
    )
    db.add(asset)
    db.commit()
    db.refresh(asset)
    return asset

# Work Orders
@router.get("/work-orders")
def list_work_orders(db: Session = Depends(get_db)):
    return db.query(FacilityWorkOrder).all()

@router.post("/work-orders")
def create_work_order(data: WorkOrderCreate, db: Session = Depends(get_db)):
    wo_id = f"WO-{uuid.uuid4().hex[:6].upper()}"
    code = f"WO-FAC-{uuid.uuid4().hex[:4].upper()}"
    asset = db.query(FacilityAsset).filter(FacilityAsset.id == data.assetId).first()
    
    wo = FacilityWorkOrder(
        id=wo_id,
        work_order_code=code,
        asset_id=data.assetId,
        asset_name=asset.name if asset else "Facility Asset",
        asset_code=asset.asset_code if asset else "AST-000",
        location=f"{asset.room if asset else 'Common Area'}, Meghna Tower",
        building_id="BLD-004",
        issue=data.issue,
        description=data.description,
        priority=data.priority,
        maintenance_type=data.maintenanceType,
        requested_by="Facility Control Center",
        assigned_department="Facility Maintenance",
        assigned_technician=data.assignedTechnician,
        vendor_name=data.vendorName,
        scheduled_date="2026-08-17",
        due_date="2026-08-18",
        sla_minutes=240,
        status="Assigned",
        estimated_cost=data.estimatedCost or 5000.0,
    )
    db.add(wo)
    db.commit()
    db.refresh(wo)
    return wo

# Utility Meters & Readings
@router.get("/utilities/meters")
def list_meters(db: Session = Depends(get_db)):
    return db.query(UtilityMeter).all()

@router.post("/utilities/readings")
def record_reading(data: UtilityReadingCreate, db: Session = Depends(get_db)):
    meter = db.query(UtilityMeter).filter(UtilityMeter.id == data.meterId).first()
    if not meter:
        raise HTTPException(status_code=404, detail="Meter not found")
    
    consumption = max(0.0, data.currentReading - meter.current_reading)
    cost = consumption * meter.rate_per_unit
    
    reading = UtilityReading(
        id=f"RDG-{uuid.uuid4().hex[:6].upper()}",
        meter_id=data.meterId,
        meter_code=meter.meter_code,
        utility_type=meter.utility_type,
        building_name=meter.building_name,
        previous_reading=meter.current_reading,
        current_reading=data.currentReading,
        consumption=consumption,
        unit=meter.unit,
        reading_date="2026-08-17",
        reader_name=data.readerName,
        reading_source="Manual",
        cost_bdt=cost,
        is_anomaly=consumption > 5000
    )
    meter.previous_reading = meter.current_reading
    meter.current_reading = data.currentReading
    
    db.add(reading)
    db.commit()
    db.refresh(reading)
    return reading
